modules/core/04-channel/types/upgrade.go

```diff
@@ -119,8 +119,17 @@ func (u *UpgradeError) GetErrorReceipt() ErrorReceipt {
 	}
 }
 
-// IsUpgradeError returns true if err is of type UpgradeError, otherwise false.
+// IsUpgradeError returns true if err is of type UpgradeError or contained
+// in the error chain of err and false otherwise.
 func IsUpgradeError(err error) bool {
-	_, ok := err.(*UpgradeError)
-	return ok
+	for {
+		_, ok := err.(*UpgradeError)
+		if ok {
+			return true
+		}
+
+		if err = errors.Unwrap(err); err == nil {
+			return false
+		}
+	}
```

Comment:
the presence of this function, means we can actually update the `Unwrap` fn of the `UpgradeError` type, I believe it can just be

```golang
// Unwrap returns the base error that caused the upgrade to fail.
func (u *UpgradeError) Unwrap() error {
	return errors.Unwrap(u.err)
}
```
And the corresponding test can be updated to 

```golang
// TestUpgradeErrorUnwrap tests that the underlying error is not modified when Unwrap is called.
func (suite *TypesTestSuite) TestUpgradeErrorUnwrap() {
	baseUnderlyingError := errorsmod.Wrap(types.ErrInvalidChannel, "base error")
	wrappedErr := errorsmod.Wrap(baseUnderlyingError, "wrapped error")
	upgradeError := types.NewUpgradeError(1, wrappedErr)

	originalUpgradeError := upgradeError.Error()
	unWrapped := errors.Unwrap(upgradeError)

	postUnwrapUpgradeError := upgradeError.Error()

	suite.Require().Equal(baseUnderlyingError, unWrapped, "unwrapped error was not equal to base underlying error")
	suite.Require().Equal(originalUpgradeError, postUnwrapUpgradeError, "original error was modified when unwrapped")
}
```

I think I like this better since it becomes more of a direct passthrough.

Reply:
this would return `nil` on unwrapping a simple wrapped err no? `UpgradeError { err: types.ErrInvalidChannel }` would call `errors.Unwrap(types.ErrInvalidChannel)` which returns `nil` yea?

Reply:
Yeah, `nil` is returned if `err` is just a single unwrapped error in that case